File: mac/Sources/App/CalDav.swift
import Foundation

/// CalDAV client for the Mac side of the bridge.
///
/// URLSession allows arbitrary HTTP methods, so unlike the Android side this
/// needs no third-party HTTP library.
struct CalDav {
    struct Item {
        var uid: String
        var href: String
        var etag: String?
        var title: String
        var notes: String?
        var due: Date?
        var allDay: Bool
        var completed: Bool

        /// Everything that matters for change detection, in one comparable string.
        var fingerprint: String {
            // Rounded to the minute and trimmed: the two sides store the same
        // reminder with different precision and whitespace.
        let minute = due.map { Int($0.timeIntervalSince1970 / 60) } ?? 0
        let note = (notes ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        return "\(title.trimmingCharacters(in: .whitespacesAndNewlines))|\(note)|\(minute)|\(allDay)|\(completed)"
        }
    }

    let user: String
    let password: String
    var root = "https://caldav.icloud.com"

    private var auth: String {
        "Basic " + Data("\(user):\(password)".utf8).base64EncodedString()
    }

    /// URLSession rewrites the method when it follows a redirect, so a REPORT
    /// arrives at iCloud's partition host as a GET and comes back 403. This
    /// session refuses to follow; `send` replays the request itself.
    private static let session: URLSession = {
        URLSession(configuration: .default, delegate: NoFollow(), delegateQueue: nil)
    }()

    private final class NoFollow: NSObject, URLSessionTaskDelegate {
        func urlSession(_ session: URLSession, task: URLSessionTask,
                        willPerformHTTPRedirection response: HTTPURLResponse,
                        newRequest request: URLRequest,
                        completionHandler: @escaping (URLRequest?) -> Void) {
            completionHandler(nil)
        }
    }

    /// Replays a request across redirects with its method and body intact.
    private func send(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
        var req = request
        for _ in 0..<5 {
            let (data, resp) = try await Self.session.data(for: req)
            guard let http = resp as? HTTPURLResponse else {
                throw BridgeError.badRequest("No HTTP response from iCloud.")
            }
            let redirect = [301, 302, 303, 307, 308].contains(http.statusCode)
            guard redirect, let location = http.value(forHTTPHeaderField: "Location"),
                  let next = URL(string: location, relativeTo: req.url) else {
                return (data, http)
            }
            req.url = next        // same method, same body, same headers
        }
        throw BridgeError.badRequest("Too many redirects from iCloud.")
    }

    // MARK: - Discovery

    /// Every collection advertising VTODO support. An iCloud home also holds
    /// inbox/outbox/shared collections that answer a REPORT with 403, so the
    /// caller has to try them in turn rather than trusting the first.
    func taskCollections() async throws -> [(href: String, name: String)] {
        let base = root.hasSuffix("/") ? root : root + "/"

        let rootXML = try await dav("PROPFIND", base, depth: 0,
            body: "<d:propfind xmlns:d=\"DAV:\"><d:prop><d:current-user-principal/></d:prop></d:propfind>")
        guard let principal = DavXML.parse(rootXML).first(where: { $0["principal-href"] != nil })?["principal-href"] else {
            throw BridgeError.badRequest("iCloud returned no principal. Check the Apple ID and app-specific password.")
        }

        let homeXML = try await dav("PROPFIND", try url(base, principal), depth: 0,
            body: "<d:propfind xmlns:d=\"DAV:\" xmlns:c=\"urn:ietf:params:xml:ns:caldav\"><d:prop><c:calendar-home-set/></d:prop></d:propfind>")
        guard let home = DavXML.parse(homeXML).first(where: { $0["home-href"] != nil })?["home-href"] else {
            throw BridgeError.badRequest("iCloud returned no calendar home.")
        }

        let homeURL = try url(base, home)
        let listXML = try await dav("PROPFIND", homeURL, depth: 1,
            body: "<d:propfind xmlns:d=\"DAV:\" xmlns:c=\"urn:ietf:params:xml:ns:caldav\"><d:prop><d:displayname/><d:resourcetype/><c:supported-calendar-component-set/></d:prop></d:propfind>")

        var found: [(href: String, name: String)] = []
        for block in DavXML.parse(listXML) where block["vtodo"] == "1" {
            guard let href = block["href"] else { continue }
            found.append((try url(homeURL, href), block["displayname"] ?? "Reminders"))
        }
        guard !found.isEmpty else {
            throw BridgeError.badRequest("No task collection in this iCloud account.")
        }
        return found
    }

    // MARK: - Read

    func items(in collection: String) async throws -> [Item] {
        let xml = try await dav("REPORT", collection, depth: 1, body:
            "<c:calendar-query xmlns:d=\"DAV:\" xmlns:c=\"urn:ietf:params:xml:ns:caldav\">"
            + "<d:prop><d:getetag/><c:calendar-data/></d:prop>"
            + "<c:filter><c:comp-filter name=\"VCALENDAR\">"
            + "<c:comp-filter name=\"VTODO\"/></c:comp-filter></c:filter></c:calendar-query>")

        return try DavXML.parse(xml).compactMap { block in
            guard let data = block["calendar-data"], data.contains("VTODO"),
                  let href = block["href"] else { return nil }
            return Ics.parse(data, href: try url(collection, href), etag: block["getetag"])
        }
    }

    // MARK: - Write

    @discardableResult
    func put(_ item: Item, in collection: String) async throws -> String {
        let href = item.href.isEmpty
            ? collection + (collection.hasSuffix("/") ? "" : "/") + item.uid + ".ics"
            : item.href
        guard let target = URL(string: href) else {
            throw BridgeError.badRequest("Bad task address: \(href)")
        }
        var req = URLRequest(url: target)
        req.httpMethod = "PUT"
        req.setValue(auth, forHTTPHeaderField: "Authorization")
        req.setValue("text/calendar; charset=utf-8", forHTTPHeaderField: "Content-Type")
        if let etag = item.etag { req.setValue(etag, forHTTPHeaderField: "If-Match") }
        req.httpBody = Data(Ics.write(item).utf8)

        let (_, http) = try await send(req)
        if !(200..<300).contains(http.statusCode), http.statusCode != 412 {
            throw BridgeError.badRequest("PUT failed: HTTP \(http.statusCode)")
        }
        return href
    }

    func delete(href: String) async throws {
        guard let target = URL(string: href) else { return }
        var req = URLRequest(url: target)
        req.httpMethod = "DELETE"
        req.setValue(auth, forHTTPHeaderField: "Authorization")
        _ = try? await send(req)
    }

    // MARK: - Plumbing

    private func dav(_ method: String, _ urlString: String, depth: Int, body: String) async throws -> String {
        guard let target = URL(string: urlString) else {
            throw BridgeError.badRequest("Bad server address: \(urlString)")
        }
        var req = URLRequest(url: target)
        req.httpMethod = method
        req.setValue(auth, forHTTPHeaderField: "Authorization")
        req.setValue("\(depth)", forHTTPHeaderField: "Depth")
        req.setValue("application/xml; charset=utf-8", forHTTPHeaderField: "Content-Type")
        req.httpBody = Data(body.utf8)

        let (data, http) = try await send(req)
        if http.statusCode == 401 {
            throw BridgeError.badRequest("iCloud rejected the password. Use an app-specific one from appleid.apple.com.")
        }
        if http.statusCode != 207, !(200..<300).contains(http.statusCode) {
            throw BridgeError.badRequest("HTTP \(http.statusCode) on \(method)")
        }
        return String(data: data, encoding: .utf8) ?? ""
    }

    /// Resolves a possibly-relative href, failing loudly instead of force-unwrapping.
    private func url(_ base: String, _ href: String) throws -> String {
        let trimmed = href.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { throw BridgeError.badRequest("iCloud returned an empty address.") }
        if trimmed.hasPrefix("http") { return trimmed }
        guard let baseURL = URL(string: base),
              let resolved = URL(string: trimmed, relativeTo: baseURL)?.absoluteString else {
            throw BridgeError.badRequest("Could not resolve \(trimmed)")
        }
        return resolved
    }
}

/// A real XML parser for 207 Multi-Status. The previous hand-rolled string
/// slicing produced an href that wasn't a URL at all, which crashed the app.
final class DavXML: NSObject, XMLParserDelegate {
    private var blocks: [[String: String]] = []
    private var current: [String: String]?
    private var stack: [String] = []
    private var text = ""

    static func parse(_ xml: String) -> [[String: String]] {
        let handler = DavXML()
        let parser = XMLParser(data: Data(xml.utf8))
        parser.delegate = handler
        parser.parse()
        return handler.blocks
    }

    private func local(_ name: String) -> String {
        (name.contains(":") ? String(name.split(separator: ":").last!) : name).lowercased()
    }

    func parser(_ p: XMLParser, didStartElement name: String, namespaceURI: String?,
                qualifiedName: String?, attributes attrs: [String: String]) {
        let n = local(name)
        stack.append(n)
        text = ""
        if n == "response" { current = [:] }
        if n == "comp", attrs["name"]?.uppercased() == "VTODO" { current?["vtodo"] = "1" }
    }

    func parser(_ p: XMLParser, foundCharacters string: String) { text += string }

    func parser(_ p: XMLParser, didEndElement name: String, namespaceURI: String?,
                qualifiedName: String?) {
        let n = local(name)
        let value = text.trimmingCharacters(in: .whitespacesAndNewlines)

        if n == "href" {
            // An href means different things depending on what encloses it.
            if stack.contains("current-user-principal") {
                current?["principal-href"] = value
            } else if stack.contains("calendar-home-set") {
                current?["home-href"] = value
            } else if current?["href"] == nil {
                current?["href"] = value
            }
        } else if ["displayname", "getetag", "calendar-data"].contains(n), !value.isEmpty {
            current?[n] = value
        } else if n == "response", let block = current {
            blocks.append(block)
            current = nil
        }
        if !stack.isEmpty { stack.removeLast() }
        text = ""
    }
}

File: android/app/src/main/java/com/saikiran/reminderssync/CalDav.java
package com.saikiran.reminderssync;

import android.util.Xml;

import org.xmlpull.v1.XmlPullParser;

import java.io.IOException;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

import okhttp3.Credentials;
import okhttp3.HttpUrl;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

/**
 * Just enough CalDAV for Apple Reminders.
 *
 * iCloud has no public API for Reminders, but it does expose them as CalDAV
 * VTODO collections. Apple doesn't document this for third parties; it has been
 * stable for years, and an app-specific password is what gets you in — the
 * account password is refused outright once two-factor auth is on.
 */
public class CalDav {
    public static final String ROOT = "https://caldav.icloud.com";
    private static final MediaType XML = MediaType.parse("application/xml; charset=utf-8");
    private static final MediaType ICAL = MediaType.parse("text/calendar; charset=utf-8");

    public static class Collection {
        public String href, name;
        public Collection(String href, String name) { this.href = href; this.name = name; }
    }

    private final OkHttpClient client;
    private final String auth;

    private final String root;

    public CalDav(String user, String appPassword) { this(user, appPassword, ROOT); }

    public CalDav(String user, String appPassword, String server) {
        this.root = (server == null || server.trim().isEmpty()) ? ROOT : server.trim();
        this.auth = Credentials.basic(user, appPassword);
        this.client = new OkHttpClient.Builder()
                .connectTimeout(20, TimeUnit.SECONDS)
                .readTimeout(45, TimeUnit.SECONDS)
                // Redirects are followed by hand below. OkHttp keeps the method
                // for PROPFIND but turns a redirected REPORT into a GET, which
                // iCloud answers with 403 — the request never arrives as a REPORT.
                .followRedirects(false)
                .followSslRedirects(false)
                .build();
    }

    // MARK: - Discovery

    /** Root → principal → calendar-home-set → the collections that hold VTODOs. */
    public List<Collection> lists() throws IOException {
        String base = root.endsWith("/") ? root : root + "/";
        String principal = firstHref(propfind(base, 0,
                "<d:propfind xmlns:d=\"DAV:\"><d:prop><d:current-user-principal/></d:prop></d:propfind>"),
                "current-user-principal");
        if (principal == null) {
            throw new IOException("No principal at " + root
                    + " — check the server address and the username.");
        }

        String home = firstHref(propfind(resolve(base, principal), 0,
                "<d:propfind xmlns:d=\"DAV:\" xmlns:c=\"urn:ietf:params:xml:ns:caldav\">"
                        + "<d:prop><c:calendar-home-set/></d:prop></d:propfind>"),
                "calendar-home-set");
        if (home == null) throw new IOException("Could not find your calendar home.");

        String homeUrl = resolve(base, home);
        String body = propfind(homeUrl, 1,
                "<d:propfind xmlns:d=\"DAV:\" xmlns:c=\"urn:ietf:params:xml:ns:caldav\">"
                        + "<d:prop><d:displayname/><d:resourcetype/>"
                        + "<c:supported-calendar-component-set/></d:prop></d:propfind>");

        List<Collection> out = new ArrayList<>();
        for (Map<String, String> r : responses(body)) {
            // Reminders live in collections that accept VTODO; calendars accept VEVENT.
            if (!"1".equals(r.get("vtodo"))) continue;
            String href = r.get("href");
            if (href == null) continue;
            String name = r.get("displayname");
            out.add(new Collection(resolve(homeUrl, href),
                                   name == null || name.isEmpty() ? "Reminders" : name));
        }
        return out;
    }

    // MARK: - Reading

    public List<Task> tasks(Collection c) throws IOException {
        String body = report(c.href,
                "<c:calendar-query xmlns:d=\"DAV:\" xmlns:c=\"urn:ietf:params:xml:ns:caldav\">"
                        + "<d:prop><d:getetag/><c:calendar-data/></d:prop>"
                        + "<c:filter><c:comp-filter name=\"VCALENDAR\">"
                        + "<c:comp-filter name=\"VTODO\"/></c:comp-filter></c:filter>"
                        + "</c:calendar-query>");

        List<Task> out = new ArrayList<>();
        for (Map<String, String> r : responses(body)) {
            String data = r.get("calendar-data");
            if (data == null || !data.contains("VTODO")) continue;
            out.add(Ics.parse(data, resolve(c.href, r.get("href")),
                              r.get("getetag"), c.href, c.name));
        }
        return out;
    }

    // MARK: - Writing

    /** Creates or replaces the .ics for this task. */
    public void save(Task t) throws IOException {
        boolean isNew = t.href == null;
        if (isNew) {
            if (t.uid == null) t.uid = UUID.randomUUID().toString().toUpperCase();
            t.href = t.listId + (t.listId.endsWith("/") ? "" : "/") + t.uid + ".ics";
        }
        Request.Builder b = new Request.Builder()
                .url(t.href)
                .header("Authorization", auth)
                .put(RequestBody.create(Ics.write(t), ICAL));
        if (isNew) b.header("If-None-Match", "*");
        else if (t.etag != null) b.header("If-Match", t.etag);

        try (Response res = follow(b, t.href)) {
            // A 412 means the server copy moved on; the next sync will reconcile.
            if (!res.isSuccessful() && res.code() != 412) {
                throw new IOException("Save failed: HTTP " + res.code());
            }
            String etag = res.header("ETag");
            if (etag != null) t.etag = etag;
        }
    }

    public void delete(Task t) throws IOException {
        if (t.href == null) return;
        Request.Builder b = new Request.Builder()
                .url(t.href).header("Authorization", auth).delete();
        if (t.etag != null) b.header("If-Match", t.etag);
        try (Response res = follow(b, t.href)) {
            if (!res.isSuccessful() && res.code() != 404) {
                throw new IOException("Delete failed: HTTP " + res.code());
            }
        }
    }

    // MARK: - Plumbing

    /** Replays a request against redirects without losing its method or body. */
    private Response follow(Request.Builder builder, String url) throws IOException {
        String target = url;
        Response res = null;
        for (int hop = 0; hop < 5; hop++) {
            if (res != null) res.close();
            res = client.newCall(builder.url(target).build()).execute();
            String location = res.header("Location");
            if (!isRedirect(res.code()) || location == null) return res;
            target = resolve(target, location);
        }
        return res;
    }

    private String propfind(String url, int depth, String body) throws IOException {
        return dav("PROPFIND", url, String.valueOf(depth), body);
    }

    private String report(String url, String body) throws IOException {
        return dav("REPORT", url, "1", body);
    }

    private String dav(String method, String url, String depth, String body) throws IOException {
        String target = url;
        for (int hop = 0; hop < 5; hop++) {
            Request req = new Request.Builder()
                    .url(target)
                    .header("Authorization", auth)
                    .header("Depth", depth)
                    .header("Content-Type", "application/xml; charset=utf-8")
                    .method(method, RequestBody.create(body, XML))
                    .build();
            try (Response res = client.newCall(req).execute()) {
                String location = res.header("Location");
                if (isRedirect(res.code()) && location != null) {
                    target = resolve(target, location);   // same method, same body
                    continue;
                }
                String text = res.body() == null ? "" : res.body().string();
                if (res.code() == 401) {
                    throw new IOException(root.contains("icloud")
                            ? "iCloud rejected the sign-in. Use an app-specific password from "
                              + "appleid.apple.com, not your Apple ID password."
                            : "Server rejected the username or password.");
                }
                if (res.code() != 207 && !res.isSuccessful()) {
                    throw new IOException("HTTP " + res.code() + " on " + method + " "
                            + shortPath(target));
                }
                return text;
            }
        }
        throw new IOException("Too many redirects from iCloud");
    }

    private static boolean isRedirect(int code) {
        return code == 301 || code == 302 || code == 303 || code == 307 || code == 308;
    }

    /** Keeps error messages readable without dumping the whole account URL. */
    private static String shortPath(String url) {
        HttpUrl u = HttpUrl.parse(url);
        if (u == null) return url;
        List<String> seg = u.pathSegments();
        return seg.isEmpty() ? "/" : "…/" + seg.get(seg.size() - 1);
    }

    private static String resolve(String base, String href) {
        if (href == null) return null;
        if (href.startsWith("http")) return href;
        HttpUrl b = HttpUrl.parse(base);
        if (b == null) return href;
        HttpUrl r = b.resolve(href);
        return r == null ? href : r.toString();
    }

    /**
     * Pulls the first href nested inside the named element. DAV puts the answer
     * in a child href rather than as text on the property itself.
     */
    private static String firstHref(String xml, String parent) {
        try {
            XmlPullParser p = Xml.newPullParser();
            p.setInput(new StringReader(xml));
            boolean inside = false;
            for (int e = p.getEventType(); e != XmlPullParser.END_DOCUMENT; e = p.next()) {
                if (e == XmlPullParser.START_TAG) {
                    if (local(p.getName()).equals(parent)) inside = true;
                    else if (inside && local(p.getName()).equals("href")) return p.nextText().trim();
                } else if (e == XmlPullParser.END_TAG && local(p.getName()).equals(parent)) {
                    inside = false;
                }
            }
        } catch (Exception ignored) { }
        return null;
    }

    /** Flattens a 207 Multi-Status into one map per &lt;response&gt;. */
    private static List<Map<String, String>> responses(String xml) {
        List<Map<String, String>> out = new ArrayList<>();
        try {
            XmlPullParser p = Xml.newPullParser();
            p.setInput(new StringReader(xml));
            Map<String, String> current = null;
            for (int e = p.getEventType(); e != XmlPullParser.END_DOCUMENT; e = p.next()) {
                if (e == XmlPullParser.START_TAG) {
                    String name = local(p.getName());
                    if (name.equals("response")) {
                        current = new HashMap<>();
                    } else if (current != null) {
                        switch (name) {
                            case "href":
                                if (!current.containsKey("href")) current.put("href", p.nextText().trim());
                                break;
                            case "displayname":   current.put("displayname", safeText(p)); break;
                            case "getetag":       current.put("getetag", safeText(p)); break;
                            case "calendar-data": current.put("calendar-data", safeText(p)); break;
                            case "comp":
                                if ("VTODO".equalsIgnoreCase(p.getAttributeValue(null, "name"))) {
                                    current.put("vtodo", "1");
                                }
                                break;
                        }
                    }
                } else if (e == XmlPullParser.END_TAG
                        && local(p.getName()).equals("response") && current != null) {
                    out.add(current);
                    current = null;
                }
            }
        } catch (Exception ignored) { }
        return out;
    }

    private static String safeText(XmlPullParser p) {
        try { return p.nextText(); } catch (Exception e) { return ""; }
    }

    /** Strips the namespace prefix so DAV:, caldav: and Apple's own all match. */
    private static String local(String name) {
        int i = name.indexOf(':');
        return (i >= 0 ? name.substring(i + 1) : name).toLowerCase();
    }
}
